polyfill/lib/duration.mjs

```diff
@@ -268,13 +271,76 @@ export class Duration {
     const maximum = maximumIncrements[smallestUnit];
     if (maximum !== undefined) ES.ValidateTemporalRoundingIncrement(roundingIncrement, maximum, false);
 
+    let zonedRelativeTo = ES.IsTemporalZonedDateTime(relativeTo) ? relativeTo : undefined;
+    let plainRelativeTo = ES.IsTemporalDate(relativeTo) ? relativeTo : undefined;
+
+    const roundingGranularityIsNoop = smallestUnit === 'nanosecond' && roundingIncrement === 1;
+    const balancingRequested = largestUnit !== existingLargestUnit;
+    const calendarUnitsPresent = years !== 0 || months !== 0 || weeks !== 0;
+    const timeUnitsOverflowWillOccur =
+      minutes >= 60 || seconds >= 60 || milliseconds >= 1000 || microseconds >= 1000 || nanoseconds >= 1000;
+    const hoursToDaysConversionMayOccur = (days !== 0 && zonedRelativeTo) || hours >= 24;
```

Comment:
The duration doesn't have to be fully balanced here, but we already can't take the optimization if any time unit would have to be balanced (see line above, `timeUnitsOverflowWillOccur`) so that includes if there are >24 hours worth of them that would have to be balanced into days.

I was unsure if I'd taken into account a duration of more than 23 hours and less than 24, rounded relative to a ZonedDateTime in a 23-hour day. But this covers that situation as well. Hours is the largest nonzero unit, so the hours-to-days conversion happens only if you specify a largestUnit of `'days'` or larger. In that case, you get `balancingRequested`, so the optimization isn't taken.

And if you try to round a duration of 1 day and 23:30 relative to a ZonedDateTime in the day before a 23-hour day, you get `hoursToDaysConversionMayOccur` because days ≠ 0 and `relativeTo` is zoned.

So I think this is covered.